esp-hal/src/uart.rs

```diff
@@ -1745,68 +1818,129 @@ impl Drop for UartTxFuture {
         // Although the isr disables the interrupt that occurred directly, we need to
         // disable the other interrupts (= the ones that did not occur), as
         // soon as this future goes out of scope.
-        self.enable_listen(false);
+        self.uart.enable_listen_tx(self.events, false);
     }
 }
 
 impl Uart<'_, Async> {
-    /// Asynchronously reads data from the UART receive buffer into the
-    /// provided buffer.
+    /// Read data asynchronously.
+    ///
+    /// This function reads data from the UART receive buffer into the
+    /// provided buffer. If the buffer is empty, the function waits
+    /// asynchronously for data to become available, or for an error to occur.
+    ///
+    /// The function returns the number of bytes read into the buffer. This may
+    /// be less than the length of the buffer.
+    ///
+    /// Note that this function may ignore the `rx_fifo_full_threshold` setting
+    /// to ensure that it does not wait for more data than the buffer can hold.
+    ///
+    /// Upon an error, the function returns immediately and the contents of the
+    /// internal FIFO are not modified.
+    ///
+    /// ## Cancellation
+    ///
+    /// This function is cancellation safe.
     pub async fn read_async(&mut self, buf: &mut [u8]) -> Result<usize, RxError> {
         self.rx.read_async(buf).await
     }
 
-    /// Asynchronously writes data to the UART transmit buffer.
+    /// Fill buffer asynchronously.
+    ///
+    /// This function reads data from the UART receive buffer into the
+    /// provided buffer. If the buffer is empty, the function waits
+    /// asynchronously for data to become available, or for an error to occur.
+    ///
+    /// Note that this function may ignore the `rx_fifo_full_threshold` setting
+    /// to ensure that it does not wait for more data than the buffer can hold.
+    ///
+    /// ## Cancellation
+    ///
+    /// This function is **not** cancellation safe. If the future is dropped
+    /// before it resolves, or if an error occurs during the read operation,
+    /// previously read data may be lost.
+    #[instability::unstable]
+    pub async fn read_exact_async(&mut self, buf: &mut [u8]) -> Result<(), RxError> {
```

Comment:
There is a couple of places in the uart tests where I've emulated this function with a loop in https://github.com/esp-rs/esp-hal/pull/3137 could you swap those out with this please?

Reply:
You didn't seem to do anything related to `async` functions in #3137 🤔 Note that I'm not implementing a blocking `read_exact`.

Reply:
Oops yeah it was for the blocking tests, see https://github.com/MabezDev/esp-hal/blob/633cfbc65430ccf12f110abb1f73c5c3717a3df1/hil-test/tests/uart.rs#L84-L91

Do we have an equivalent `read_exact` for blocking?

Reply:
Not directly, but we have it through embedded_io::Read I believe.